### PR TITLE
Viewport scaling

### PR DESCRIPTION
**Pull request by PatrickvL**
This PR brings these (related) improvements:

* a fix for #1096 - half-sized backbuffer display modes
* a hack configuration option that does :
  * the main window is resized according to the video resolution configuration
  * the D3D viewport is scaled according to the main window dimensions

Do note, that higher resolutions than 640x480 could show strange results; Some titles don't stretch their background images, but render 3D in the correct scale - but as the Burnout car-selection screen shows, in this case 3D is only drawn in the part that has a background image. Other titles (like Battlestar Galactica) scale perfectly. Some XDK samples fill the entire window correctly, but the text labels are still at the original (unscaled) positions.

### CODE REVIEW

**Comment by ESPPIRAL** [comment]
bandicam 2018-04-29 15-35-53-248

**Comment by ESPPIRAL** [comment]
Shenmue 2 still only renders half the screen.

bandicam 2018-04-29 15-41-28-073

**Comment by ESPPIRAL** [comment]
Barbarian renders to a black screen, you can actually hear the game sound.

bandicam 2018-04-29 15-46-13-821

On master it renders just fine.
bandicam 2018-04-29 15-47-45-026

**Comment by ESPPIRAL** [comment]
Jet set Radio Future, It crashes upon loading a savedgame, it don't crash on master.

bandicam 2018-04-29 15-56-45-260

**Comment by ESPPIRAL** [comment]
Taz Wanted, similar to SHenmue but it renders on the left.

bandicam 2018-04-29 16-06-33-905

**Comment by ESPPIRAL** [comment]
Same with Turok

bandicam 2018-04-29 16-08-59-480

**Comment by ESPPIRAL** [comment]
Scrapland Scales 2D images
bandicam 2018-04-29 16-11-21-883

But during gameplay it zooms in to the right corner.

bandicam 2018-04-29 16-12-05-957

**Comment by ESPPIRAL** [comment]
CRASH WARTH OF CORTEX does the same.

bandicam 2018-04-29 16-35-07-104

**Comment by PatrickvL** [comment]
From discord :
what happens if you run these titles with a configured resolution of 640x480?

Crash looks as allways (correct)
Shenmue Still renders on the top half
Jet set radio still crashes
Doa 3 looks normal (correct)
scrapland looks ok too

**Comment by ESPPIRAL** [comment]
Crash looks as allways

**Comment by ESPPIRAL** [comment]
Shenmue Still renders on the top half



Doa 3 looks normal



scrapland looks ok too




JSRF crashes at the same point when you load your savegame an it should take yo to the charecter selection screen

**Comment by Margen67** [comment]
@PatrickvL Any chance for anti aliasing or anisotropic filtering options, or is that beyond the scope of this PR?

**Comment by bhodg22** [comment]
Namco Museum works well by stretching the image to fit
image

Rallisport demo only displays in the top left corner
image

As mentioned previously JSRF works well but crashes when loading singleplayer
image

Crash Twinsanity kinds works but UI/hud elements are at the top left, hard to see
image

**Comment by HyperScorpio** [comment]
I'm not using HD Scales for now on, we should be sticked throw the original resolution cxbx as provide us first to fix all games and then they try to move on in to the next step. (I think)

**Comment by RadWolfie** [comment]
LGTM

**Comment by Margen67** [comment]
Tony Hawk Pro Skater 2X boots without the viewport hack enabled, but now shows a new error when it is enabled.
Release:
thps2x
Debug: 
thps2xdebug

Debug log: [KrnlDebug.txt](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/files/1968643/KrnlDebug.txt)


Spongebob - Battle for Bikini Bottom shows this error:
bfbberror
With the viewport hack disabled you can continue, but with it enabled it crashes with this error.
Release:
bfbberror2release
Debug:
bfbberror2debug
Debug log:
[KrnlDebug.zip](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/files/1968698/KrnlDebug.zip)

**Comment by PatrickvL** [comment]
@Margen67 thanks for those details - is any of the errors present in master too?

**Comment by Margen67** [comment]
@PatrickvL No.

**Comment by PatrickvL** [comment]
But when the hack is disabled, behaviour is identical to master, right?

(Just asking to ascertain if it's a regression or not)

**Comment by Margen67** [comment]
@PatrickvL The games do boot with the hack disabled.

File: src/Cxbx/WndMain.cpp [hunk 1]
```diff
@@ -113,6 +113,52 @@ void WndMain::InitializeSettings() {
...
+		}
+	}
+
+	// TODO : Acknowledge DPI scaling here
+	GetWindowRect(GetDesktopWindow(), &desktopRect);
+
+	// Limit width/height to desktop resolution
+	int dWidth = desktopRect.right - desktopRect.left;
+	int dHeight = desktopRect.bottom - desktopRect.top;
+	if (m_w > dWidth)
```
Comment (LukeUsher) [inline src/Cxbx/WndMain.cpp]: A minor thing, but please use braces around even single-line if statements, for the purposes of clarity/readability & maintenance.

**Comment by NNX3x** [comment]
NBA LIVE 2005 and Prince of Persia 2T show this alert.

2t

**Pull request merged**